### PR TITLE
[perf] Rollup EnvoyFilter patches

### PR DESCRIPTION
**Pull request by stevenctl**
Screen Shot 2020-06-21 at 5 56 17 PM

[download svg (zip)](https://github.com/istio/istio/files/4810686/profile.zip)

While experimenting with profiling pilot under different scenarios, I tried profiling pilot while running [istio/tools/perf/load/setup_large_test.sh](https://github.com/istio/tools/blob/master/perf/load/setup_large_test.sh) with NUM=15 and saw that there were a lot (~18%) of samples for `proto.Merge` via `ApplyClusterPatches`. 

This PR optimizes `ApplyXPatches` to make fewer `proto.Merge` calls by rolling up MERGE patches by their matcher.

<table class='benchstat oldnew'>
<tr class='configs'><th><th>base.txt<th>428064f823e166a5667a59cb8c722db325224053.txt

<tbody>
<tr><th><th colspan='2' class='metric'>time/op<th>delta
<tr class='unchanged'><td>RouteGeneration/gateways-4<td>73.9ms ± 2%<td>72.0ms ± 3%<td class='nodelta'>~<td class='note'>(p=0.095 n=5&#43;5)
<tr class='unchanged'><td>RouteGeneration/gateways-shared-4<td>289ms ± 4%<td>286ms ± 3%<td class='nodelta'>~<td class='note'>(p=0.421 n=5&#43;5)
<tr class='better'><td>RouteGeneration/empty-4<td>1.32ms ± 8%<td>1.22ms ± 2%<td class='delta'>−7.86%<td class='note'>(p=0.032 n=5&#43;5)
<tr class='better'><td>RouteGeneration/tls-4<td>1.27ms ± 1%<td>1.23ms ± 1%<td class='delta'>−2.83%<td class='note'>(p=0.008 n=5&#43;5)
<tr class='unchanged'><td>RouteGeneration/telemetry-4<td>1.23ms ± 5%<td>1.26ms ± 3%<td class='nodelta'>~<td class='note'>(p=0.222 n=5&#43;5)
<tr class='worse'><td>RouteGeneration/virtualservice-4<td>2.61ms ± 1%<td>2.76ms ± 1%<td class='delta'>&#43;5.71%<td class='note'>(p=0.008 n=5&#43;5)
<tr class='unchanged'><td>ClusterGeneration/gateways-4<td>5.33ms ± 2%<td>5.34ms ± 1%<td class='nodelta'>~<td class='note'>(p=0.548 n=5&#43;5)
<tr class='unchanged'><td>ClusterGeneration/gateways-shared-4<td>5.57ms ± 1%<td>5.60ms ± 3%<td class='nodelta'>~<td class='note'>(p=0.841 n=5&#43;5)
<tr class='unchanged'><td>ClusterGeneration/empty-4<td>2.14ms ± 2%<td>2.18ms ± 6%<td class='nodelta'>~<td class='note'>(p=0.222 n=5&#43;5)
<tr class='unchanged'><td>ClusterGeneration/tls-4<td>7.23ms ± 1%<td>7.14ms ± 1%<td class='nodelta'>~<td class='note'>(p=0.095 n=5&#43;5)
<tr class='worse'><td>ClusterGeneration/telemetry-4<td>2.17ms ± 1%<td>2.25ms ± 3%<td class='delta'>&#43;3.93%<td class='note'>(p=0.008 n=5&#43;5)
<tr class='unchanged'><td>ClusterGeneration/virtualservice-4<td>526µs ± 2%<td>530µs ± 1%<td class='nodelta'>~<td class='note'>(p=0.421 n=5&#43;5)
<tr class='unchanged'><td>ListenerGeneration/gateways-4<td>18.9ms ± 0%<td>19.1ms ± 2%<td class='nodelta'>~<td class='note'>(p=0.548 n=5&#43;5)
<tr class='worse'><td>ListenerGeneration/gateways-shared-4<td>38.4µs ± 1%<td>38.9µs ± 1%<td class='delta'>&#43;1.39%<td class='note'>(p=0.008 n=5&#43;5)
<tr class='unchanged'><td>ListenerGeneration/empty-4<td>1.59ms ± 7%<td>1.63ms ± 6%<td class='nodelta'>~<td class='note'>(p=0.095 n=5&#43;5)
<tr class='worse'><td>ListenerGeneration/tls-4<td>1.50ms ± 3%<td>1.55ms ± 2%<td class='delta'>&#43;3.13%<td class='note'>(p=0.032 n=5&#43;5)
<tr class='worse'><td>ListenerGeneration/telemetry-4<td>1.71ms ± 1%<td>1.77ms ± 1%<td class='delta'>&#43;3.91%<td class='note'>(p=0.008 n=5&#43;5)
<tr class='worse'><td>ListenerGeneration/virtualservice-4<td>110µs ± 1%<td>117µs ± 9%<td class='delta'>&#43;6.49%<td class='note'>(p=0.008 n=5&#43;5)
<tr class='unchanged'><td>EndpointGeneration/1/100-4<td>337µs ± 2%<td>338µs ± 0%<td class='nodelta'>~<td class='note'>(p=0.222 n=5&#43;5)
<tr class='unchanged'><td>EndpointGeneration/10/10-4<td>128µs ± 0%<td>130µs ± 1%<td class='nodelta'>~<td class='note'>(p=0.056 n=5&#43;5)
<tr class='unchanged'><td>EndpointGeneration/100/10-4<td>1.04ms ± 1%<td>1.05ms ± 1%<td class='nodelta'>~<td class='note'>(p=0.095 n=5&#43;5)
<tr class='unchanged'><td>EndpointGeneration/1000/1-4<td>1.01ms ± 1%<td>1.02ms ± 1%<td class='nodelta'>~<td class='note'>(p=0.151 n=5&#43;5)
<tr><td>&nbsp;
</tbody>

</table>
</body>
</html>

Would it make sense to apply this rollup closer to the config store, only updating it when the configuration changes?

### CODE REVIEW

**Comment by rshriram** [comment]
this is great! thanks for profiling the code. Do you have a sense about the correctness when you do merge rollup?

**Comment by howardjohn** [comment]
I recommend https://godoc.org/golang.org/x/perf/cmd/benchstat for comparing two benchmarks

**Comment by howardjohn** [comment]
You should try out the benchmarks here: https://github.com/istio/istio/blob/b0620237499985147dc86026148a1691aa2750be/pilot/pkg/proxy/envoy/v2/bench_test.go#L196

They have a telemetry v2 test, which should give realistic impact on real world results. Looking at the filters, I *think* this will have a substantially impact. Clusters probably not as much, but the listener patches are ~10% of envoy CPU in real world clusters, so if we apply this logic there we could really improve things

**Comment by stevenctl** [comment]
I have an initial version of the rollup logic for listeners, but that benchmark doesn't seem to do much with ApplyListenerPatches – It either [returns early](https://github.com/istio/istio/blob/77d9c1040b1a56064f7e59593f53331cca6c7578/pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go#L57) with a nil `EnvoyFilterWrapper` or builds the wrapper but with 0 patches of any kind. 

I will look into making the benchmark provide some stress to `ApplyListenerPatches` or repair the existing benchmark in [pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch_test.go](https://github.com/istio/istio/blob/77d9c1040b1a56064f7e59593f53331cca6c7578/pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch_test.go) which seems to rely on some files that were moved/deleted. 

I'm also going to do another version of the rollup logic where the rollup is applied to `EnvoyFilter_LISTENER` patches first, then applied to `EnvoyFilter_FILTER_CHAIN` matches that match the listener. The current version will not roll up filter chain patches unless the patch doesn't check for a specific listener.

**Comment by stevenctl** [comment]
I was able to get the mentioned benchmark to actually exercise ApplyListenerPatches - every patch was filtered out because of the proxy version check. (See https://github.com/istio/istio/pull/24849/commits/479db3278ba77b646849404971064f945fc79c78)

Even with that, it only tries INSERT_BEFORE patches the listener path (network filters to be specific). The logic so far only helps with MERGE. I'll need to add additional patches to the benchmark, but if it's used for other comparisons that would invalidate those results. 

I may also try to reduce the number of Clone calls necessary.

**Comment by stevenctl** [comment]
I added a sort of contrived scenario for the BenchmarkListenerGeneration (telemetry) in https://github.com/istio/istio/pull/24849/commits/d0ca4a07b397509f093025b5e5b9ef05d8dc81aa with a bunch of MERGE patches for network filters, just because there weren't any merge patches to begin with.

There is some slight speedup and the performance of other scenarios is unaffected. 

<pre>
benchstat benchtest_merge_master.txt benchtest_globalcheck.txt
name                                  old time/op  new time/op  delta
RouteGeneration/empty-12               226ns ± 0%   225ns ± 0%   ~     (p=1.000 n=1+1)
RouteGeneration/telemetry-12           226ns ± 0%   228ns ± 0%   ~     (p=1.000 n=1+1)
RouteGeneration/virtualservice-12      229ns ± 0%   223ns ± 0%   ~     (p=1.000 n=1+1)
ClusterGeneration/empty-12            2.86ms ± 0%  2.95ms ± 0%   ~     (p=1.000 n=1+1)
ClusterGeneration/telemetry-12        4.90ms ± 0%  5.17ms ± 0%   ~     (p=1.000 n=1+1)
ClusterGeneration/virtualservice-12   2.87ms ± 0%  2.90ms ± 0%   ~     (p=1.000 n=1+1)
ListenerGeneration/empty-12           1.98ms ± 0%  1.99ms ± 0%   ~     (p=1.000 n=1+1)
<b>ListenerGeneration/telemetry-12       14.3ms ± 0%   5.3ms ± 0%   ~     (p=1.000 n=1+1)</b>
ListenerGeneration/virtualservice-12  2.00ms ± 0%  1.99ms ± 0%   ~     (p=1.000 n=1+1)
EndpointGeneration/100/1-12           79.2µs ± 0%  77.9µs ± 0%   ~     (p=1.000 n=1+1)
EndpointGeneration/1000/1-12           722µs ± 0%   705µs ± 0%   ~     (p=1.000 n=1+1)
EndpointGeneration/10000/1-12         7.15ms ± 0%  7.04ms ± 0%   ~     (p=1.000 n=1+1)
EndpointGeneration/100/100-12         7.80ms ± 0%  7.63ms ± 0%   ~     (p=1.000 n=1+1)
EndpointGeneration/1000/100-12        76.3ms ± 0%  76.0ms ± 0%   ~     (p=1.000 n=1+1)
EndpointGeneration/10000/100-12        1.17s ± 0%   1.15s ± 0%   ~     (p=1.000 n=1+1)
</pre>

I also consolidated the logic used for rolling up cluster patches and performance was unaffected vs the original optimization. 

<pre>
benchstat cp_master.txt cp_rollup.txt cp_globalcheck.txt
name \ time/op          cp_master.txt  cp_rollup.txt  cp_globalcheck.txt
ApplyClusterPatches-12     569ms ± 0%       6ms ± 0%            6ms ± 0%
</pre>

**Comment by howardjohn** [comment]
btw, the reason the "telemetry" test is important as is is because 99% percent of users have those filters so it impacts everyone. Other envoyfilter optimizations are useful as well but impact less people - although it probably is skewed towards large scale users as its advanced config

**Comment by howardjohn** [comment]
Also with benchstat if you run the benchmarks with `-count=N` it will give a delta and p value which can be nice, but of course takes N times longer

**Comment by stevenctl** [comment]
> this is great! thanks for profiling the code. Do you have a sense about the correctness when you do merge rollup?

The merge ordering is effectively unchanged. Existing UTs should prove that.

**Comment by stevenctl** [comment]
> btw, the reason the "telemetry" test is important as is is because 99% percent of users have those filters so it impacts everyone. Other envoyfilter optimizations are useful as well but impact less people - although it probably is skewed towards large scale users as its advanced config

It might be tough to do much for the common-case of inserting listeners. Still not 100% totally convinced we need all the `proto.Clone` calls though; 

https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go#L89

We never mutate anything after insertion. If we still want some safety we could do an upfront clone so that members of `PushContext.envoyFiltersByNamespace` are guaranteed to remain unmodified.

**Comment by stevenctl** [comment]
<pre>
name \ time/op                        listenergen_master_fixed.txt  listenergen_cpclone.txt  listenergen_noclone.txt
ListenerGeneration/empty-12                            2.11ms ± 3%              2.08ms ± 1%              2.10ms ± 1%
<b>ListenerGeneration/telemetry-12                        3.66ms ± 1%              3.25ms ± 1%              3.32ms ± 6%</b>
ListenerGeneration/virtualservice-12                   2.13ms ± 6%              2.20ms ±12%              2.19ms ±14%
</pre>



Got a ~10% save by only cloning the value in a given ConfigPatchWrapper at most once. Not cloning at all didn't help at all. 

```
benchstat listenergen_master_fixed.txt listenergen_cpclone.txt                                                                            
name                                  old time/op  new time/op  delta
ListenerGeneration/empty-12           2.11ms ± 3%  2.08ms ± 1%   -1.53%  (p=0.000 n=20+18)
ListenerGeneration/telemetry-12       3.66ms ± 1%  3.25ms ± 1%  -11.24%  (p=0.000 n=16+17)
ListenerGeneration/virtualservice-12  2.13ms ± 6%  2.20ms ±12%     ~     (p=0.228 n=18+20)
```

**Comment by howardjohn** [comment]
@stevenctl are you still working on this? Seems very useful but is marked WIP. Just going through some old PRs

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/rollup.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+		objectMatcher = cp.Match.ObjectTypes.(*networking.EnvoyFilter_EnvoyConfigObjectMatch_Cluster).Cluster
+	default:
+		return "", fmt.Errorf("unsupported object matcher in patch wrapper; got %T", cp.Match.ObjectTypes)
+	}
+
+	if objectMatcher == nil {
+		return "", fmt.Errorf("empty matcher for patch")
+	}
+
+	hash, err := proto.Marshal(objectMatcher)
```
Comment (stevenctl) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/rollup.go:124]: This is the fastest way I've found to make a key for these. I thought [deepmind/objecthash-proto](https://github.com/deepmind/objecthash-proto) might work but it's slower and has issues with 1.20+

**Comment by stevenctl** [comment]
@howardjohn as far as bench_test scenarios it doesn't affect much:

```
name                                   old time/op        new time/op        delta
RouteGeneration/gateways-12                  48.6ms ±22%        44.6ms ± 2%     ~     (p=0.095 n=5+5)
RouteGeneration/gateways-shared-12            216ms ±37%         189ms ± 5%     ~     (p=0.151 n=5+5)
RouteGeneration/empty-12                      968µs ±73%         713µs ± 3%  -26.37%  (p=0.008 n=5+5)
RouteGeneration/tls-12                        735µs ±10%         740µs ± 8%     ~     (p=0.690 n=5+5)
RouteGeneration/telemetry-12                  708µs ± 3%         719µs ± 4%     ~     (p=0.421 n=5+5)
RouteGeneration/virtualservice-12            1.55ms ± 1%        1.56ms ± 2%     ~     (p=0.548 n=5+5)
ClusterGeneration/gateways-12                3.07ms ± 2%        3.44ms ±20%  +12.20%  (p=0.016 n=5+5)
ClusterGeneration/gateways-shared-12         3.07ms ± 1%        3.07ms ± 2%     ~     (p=0.841 n=5+5)
ClusterGeneration/empty-12                   1.25ms ± 2%        1.27ms ± 1%     ~     (p=0.063 n=5+4)
ClusterGeneration/tls-12                     4.09ms ± 0%        4.47ms ±11%   +9.20%  (p=0.008 n=5+5)
ClusterGeneration/telemetry-12               1.24ms ± 0%        1.31ms ± 7%     ~     (p=0.151 n=5+5)
ClusterGeneration/virtualservice-12           303µs ± 1%         333µs ± 8%   +9.93%  (p=0.008 n=5+5)
ListenerGeneration/gateways-12               13.0ms ± 2%        14.7ms ±18%  +13.09%  (p=0.016 n=5+5)
ListenerGeneration/gateways-shared-12        27.7µs ± 4%        27.5µs ± 5%     ~     (p=0.690 n=5+5)
ListenerGeneration/empty-12                  1.16ms ± 5%        1.24ms ±24%     ~     (p=0.548 n=5+5)
ListenerGeneration/tls-12                    1.09ms ± 1%        1.19ms ± 7%   +8.26%  (p=0.008 n=5+5)
ListenerGeneration/telemetry-12              1.25ms ± 1%        1.25ms ± 1%     ~     (p=0.548 n=5+5)
ListenerGeneration/virtualservice-12         83.0µs ± 1%        84.1µs ± 4%     ~     (p=0.222 n=5+5)
EndpointGeneration/1/100-12                   236µs ±16%         224µs ± 1%     ~     (p=1.000 n=5+5)
EndpointGeneration/10/10-12                  89.6µs ±13%        87.6µs ± 1%     ~     (p=0.841 n=5+5)
EndpointGeneration/100/10-12                  737µs ±14%         719µs ± 2%     ~     (p=0.841 n=5+5)
EndpointGeneration/1000/1-12                  682µs ± 1%         703µs ± 2%   +3.06%  (p=0.008 n=5+5)

```

**Comment by stevenctl** [comment]
If we want, we can pull this in for clusters. I don't think it will do much for most people's config. Closing for now, xds caching will be a lot better.

**Pull request closed without merging**